src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs

```diff
@@ -269,24 +269,31 @@ internal int ReadCore(Span<byte> buffer)
                     break;
                 }
 
-                if (_inflater.Finished())
+                // If the stream is finished then we have a few potential cases here:
+                // 1. DeflateStream => return
+                // 2. GZipStream that is finished but may have an additional GZipStream appended => feed more input
+                // 3. GZipStream that is finished and appended with garbage => return
+                if (_inflater.Finished() && (!_inflater.IsGzipStream() || !_inflater.NeedsInput()))
```

Comment:
What happens if the underlying stream had unrelated data after the GZipStream that was going to be read by something unrelated after the first part was decompressed... are we potentially corrupting cases that today work, unrelated to multiple gzips in a single stream?

Reply:
That is currently broken today, as we will already read past the end of the valid gzipstream to fill our input buffer. Currently we just throw away the extra bytes after the first gzip data stream ends.

This change doesn't affect that case since we're still filling the same buffer. The only scenario that we would be doing something different is if that garbage happens to start with the gzip header bytes (causing us to throw).

Reply:
> That is currently broken today

Ok... and we don't plan to fix it?  Or we'll fix it separately?

Reply:
Either one. I'm not sure how we'd fix it on a non-seekable stream unless we fed input to zlib one byte at a time.